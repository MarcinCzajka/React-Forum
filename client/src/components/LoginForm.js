import React from 'react';
import { Message, Button, Form, Modal, Menu, Checkbox } from 'semantic-ui-react';
import basePath from '../api/basePath';
import UserContext from '../contexts/UserContext';
import jwt_decode from 'jwt-decode';

class LoginForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            'open': false,
            'stayLogged': false,
            'buttonName': '',
            'email': '',
            'password': '',
            'error': ''
        }
    }

    static contextType = UserContext;
    
    componentDidMount() {
        this.context.addShowLoginToContext(this.open);
    }

    login = () => {
        basePath({
            method: 'post',
            url: '/login',
            data: {
                email: this.state.email,
                password: this.state.password
            },
            withCredentials: true
        }).then(res => {
            if (res.status === 200) {
                const data = {...jwt_decode(document.cookie), ...{loggedIn: true}};
                this.context.setContextData(data);

                if(this.state.stayLogged) localStorage.setItem('token', document.cookie);

                this.setState({
                    'open': false,
                    'buttonName': 'Log out',
                    'email': '',
                    'password': '',
                    'passwordRepeat': '',
                    'error': ''
                })
            }
        }).catch(error => {
            console.log(error)
            this.setState({ error: error.response.data })
        });
    }

    logout = () => {
        localStorage.removeItem('token');
        document.cookie = 'x-auth-token=;expires=Thu, 01 Jan 1970 00:00:01 GMT;';
        
        this.context.setContextData({
            loggedIn: false,
            userName: '',
            userId: '',
            userAvatar: '',
            userEmail: '',
            userCreatedAt: ''
        })
    }

    error = () => {
        if (this.state.error) {
            return (
                <Message error
                header = 'Error'
                content = {this.state.error} />
            )
        }
    }

    close = () => {
        this.setState({open: false});
    }

    open = () => {
        this.setState({open: true});
    }

    render() {
        return (
            !this.context.loggedIn ? (
            <Modal size='tiny' trigger={<Menu.Item onClick={this.open} name="Log in" />}
                open={this.state.open}
                onClose={this.close}
            >
                <Modal.Header>Log in</Modal.Header>
                <Modal.Content style={{paddingTop: 0}}>
                    <Form size='small' onSubmit={this.login} error >
                        <div>{this.error()}</div>
                        <br></br>
                        <Form.Field required>
                            <label>Email</label>
                            <input placeholder='Email' value={this.state.email} onChange={ (e) => this.setState({ email: e.target.value })} />
                        </Form.Field>
                        <Form.Field required>
                            <label>Password</label>
                            <input 
                                type='password'
                                placeholder='Password' 
                                autoComplete = "new-password"
                                spellCheck = "false"
                                autoCapitalize = "off"
                                autoCorrect = "off"
                                value={this.state.password} 
                                onChange={(e) => this.setState({ password: e.target.value })} 
                            />
                        </Form.Field>
                        <Checkbox
                            checked={this.state.stayLogged}
                            label='Stay logged in' 
                            onChange={() => this.setState({ stayLogged: !this.state.stayLogged })} 
                        />

                        <Button style={{'marginTop':'15px'}} type='submit' fluid size='large'>Sign in</Button>
                    </Form>
                </Modal.Content>
            </Modal>
            ) : (
                <Menu.Item onClick={this.logout} name="Log out" />
            )
        )
    }

}

export default LoginForm;