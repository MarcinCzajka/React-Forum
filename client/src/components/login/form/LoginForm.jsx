import React from 'react';
import PropTypes from 'prop-types';
import { Button, Form, Modal, Checkbox } from 'semantic-ui-react';
import ErrorMessage from '../../message/ErrorMessage';
import { LocaleConsumer } from '../../../contexts/LocaleContext';

class LoginForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            email: '',
            password: '',
            stayLogged: false
        }
    }

    render() {
        const { open, onClose, handleLogin, errorMessage, showSignup } = this.props;

        return (
            <LocaleConsumer>
                {locale => (
                    <Modal 
                        size='tiny'
                        open={open}
                        onClose={onClose}
                    >
                        <Modal.Header>{locale.login.modalHeader}</Modal.Header>

                        <Modal.Content>
                            <Form 
                                error
                                size='small'
                                onSubmit={ () => { handleLogin(this.state) }}
                            >

                                <ErrorMessage message={errorMessage} />

                                <Form.Field required>
                                    <label>{locale.login.label.email}</label>
                                    <input
                                        value={this.state.email}
                                        placeholder={locale.login.label.emailPlaceholder}
                                        onChange={ (e) => this.setState({ email: e.target.value })}
                                    />
                                </Form.Field>

                                <Form.Field required>
                                    <label>{locale.login.label.password}</label>
                                    <input 
                                        type='password'
                                        placeholder={locale.login.label.passwordPlaceholder} 
                                        autoComplete = "new-password"
                                        spellCheck = "false"
                                        autoCapitalize = "off"
                                        autoCorrect = "off"
                                        value={this.state.password} 
                                        onChange={ (e) => this.setState({ password: e.target.value })} 
                                    />
                                </Form.Field>

                                <Checkbox
                                    checked={this.state.stayLogged}
                                    label={locale.login.stayLoggedIn}
                                    onChange={() => this.setState({ stayLogged: !this.state.stayLogged })} 
                                />

                                <Button style={{'marginTop':'15px'}} type='submit' fluid size='large'>{locale.login.submitButton}</Button>

                                <p>{locale.login.noAccountYet} <span className='asLink' onClick={showSignup}>{locale.login.signUpLink}</span>!</p>

                            </Form>
                        </Modal.Content>
                        
                    </Modal>
                )}
            </LocaleConsumer>
        )
    }
}

LoginForm.propTypes = {
    open: PropTypes.func.isRequired,
    onClose: PropTypes.func.isRequired,
    handleLogin: PropTypes.func.isRequired,
    errorMessage: PropTypes.string,
    showSignup: PropTypes.func.isRequired
}

export default LoginForm;