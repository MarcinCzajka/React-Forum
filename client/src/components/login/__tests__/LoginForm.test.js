import React from 'react';
import { create } from "react-test-renderer";
import { shallow } from 'enzyme';
import LoginForm from '../form/LoginForm';
import { LocaleProvider } from '../../../contexts/LocaleContext';
import locale from '../../../locale/en/dictionary.json';

//This line is needed to render semantic-ui Modal content because of Portals used in this component
jest.mock('semantic-ui-react/dist/commonjs/addons/Portal/Portal', () => ({ children }) => children);

describe('LoginForm component', () => {
    test('renders without crashing', () => {
        shallow(<LocaleProvider value={locale}><LoginForm /></LocaleProvider>);
    });

    test('match snapshot', () => {
        const component = create(<LocaleProvider value={locale}><LoginForm /></LocaleProvider>);
        expect(component).toMatchSnapshot();
    });

});