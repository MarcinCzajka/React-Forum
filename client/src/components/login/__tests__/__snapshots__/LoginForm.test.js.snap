// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component match snapshot /w user logged in 1`] = `
<a
  className="item"
  onClick={[Function]}
>
  Log Out
</a>
`;

exports[`LoginForm component match snapshot w/o user logged in 1`] = `
<div
  className="ui page modals dimmer transition visible active"
>
  <div
    className="ui tiny modal transition visible active noMargin"
    style={Object {}}
  >
    <div
      className="header"
    >
      Log in
    </div>
    <div
      className="content"
    >
      <form
        className="ui small error form"
        onSubmit={[Function]}
      >
        
        <div
          className="required field"
        >
          <label>
            Email
          </label>
          <input
            onChange={[Function]}
            placeholder="Email"
            value=""
          />
        </div>
        <div
          className="required field"
        >
          <label>
            Password
          </label>
          <input
            autoCapitalize="off"
            autoComplete="new-password"
            autoCorrect="off"
            onChange={[Function]}
            placeholder="Password"
            spellCheck="false"
            type="password"
            value=""
          />
        </div>
        <div
          className="ui checkbox"
          onChange={[Function]}
          onClick={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
        >
          <input
            checked={false}
            className="hidden"
            readOnly={true}
            tabIndex={0}
            type="checkbox"
          />
          <label>
            Stay logged in
          </label>
        </div>
        <button
          className="ui large fluid button"
          onClick={[Function]}
          style={
            Object {
              "marginTop": "15px",
            }
          }
          type="submit"
        >
          Sign in
        </button>
        <p>
          Don't have account yet? 
          <span
            className="asLink"
            onClick={[Function]}
          >
            Sign up
          </span>
          !
        </p>
      </form>
    </div>
  </div>
</div>
`;
